Source/DafnyCore/Backends/Rust/Dafny-compiler-rust.dfy

```diff
@@ -3904,33 +4190,46 @@ module {:extern "DCOMP"} DafnyToRustCompiler {
     function UpcastConversionLambda(fromType: Type, fromTpe: R.Type, toType: Type, toTpe: R.Type, typeParams: map<(R.Type, R.Type), R.Expr>): Result<R.Expr, (Type, R.Type, Type, R.Type, map<(R.Type, R.Type), R.Expr>)>
     {
       if fromTpe == toTpe then
-        Success(R.dafny_runtime.MSel("upcast_id").ApplyType([fromTpe]).Apply([]))
+        Success(R.dafny_runtime.MSel("upcast_id").AsExpr().ApplyType([fromTpe]).Apply([]))
       else if fromTpe.IsObjectOrPointer() && toTpe.IsObjectOrPointer() then
         if !toTpe.ObjectOrPointerUnderlying().DynType? then Failure((fromType, fromTpe, toType, toTpe, typeParams)) else
         var fromTpeUnderlying := fromTpe.ObjectOrPointerUnderlying();
         var toTpeUnderlying := toTpe.ObjectOrPointerUnderlying();
-        Success(R.dafny_runtime.MSel(upcast).ApplyType([fromTpeUnderlying, toTpeUnderlying]).Apply([]))
+        Success(R.dafny_runtime.MSel(upcast).AsExpr().ApplyType([fromTpeUnderlying, toTpeUnderlying]).Apply([]))
       else if (fromTpe, toTpe) in typeParams then
         Success(typeParams[(fromTpe, toTpe)])
       else if fromTpe.IsRc() && toTpe.IsRc() then
         var lambda :- UpcastConversionLambda(fromType, fromTpe.RcUnderlying(), toType, toTpe.RcUnderlying(), typeParams);
         if fromType.Arrow? then
           Success(lambda)
         else
-          Success(R.dafny_runtime.MSel("rc_coerce").Apply1(lambda))
+          Success(R.dafny_runtime.MSel("rc_coerce").AsExpr().Apply1(lambda))
       else if SameTypesButDifferentTypeParameters(fromType, fromTpe, toType, toTpe) then
-        var lambdas :- SeqResultToResultSeq(seq(|fromTpe.arguments|, i requires 0 <= i < |fromTpe.arguments| =>
-                                              UpcastConversionLambda(
-                                                fromType.resolved.typeArgs[i],
-                                                fromTpe.arguments[i],
-                                                toType.resolved.typeArgs[i],
-                                                toTpe.arguments[i],
-                                                typeParams
-                                              )));
+        var indices :=
+          if fromType.UserDefined? && fromType.resolved.kind.Datatype? then
+            Std.Collections.Seq.Filter(i =>
```

Comment:
If you think of
``` 
            Std.Collections.Seq.Filter(i requires 0 <= i < |fromTpe.arguments| =>
                                           (0 <= i < |fromType.resolved.kind.variances| ==>
                                              !fromType.resolved.kind.variances[i].Nonvariant?)
                                      , seq(|fromTpe.arguments|, i => i))
```
it will fail verification later on on:
```
seq(|indices|, j requires 0 <= j < |indices| =>
                           var i := indices[j];
                           ...
                             fromType.resolved.typeArgs[i],
                               ^^^^ îndex out of range.
```
The reason for that is that it does not know after filtering that the index is still in bounds. Not sure why. But my workaround works fine so I'll stick with that for now.

Reply:
Dang, thanks for trying anyway